File: backend/app/schemas/contract.py
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class TaskStatus(str, Enum):
    PENDING_UPLOAD = "pending_upload"
    PROCESSING = "processing"
    COMPLETED = "completed"
    NEEDS_REVIEW = "needs_review"


class EvidenceRef(BaseModel):
    id: str
    page: int
    bbox: tuple[int, int, int, int]
    text: str
    sourceType: str
    sourceId: str
    segmentIndex: int = 0
    segmentCount: int = 1
    isPrimary: bool = True
    accent: str = "cyan"


class DocumentBlock(BaseModel):
    id: str
    text: str
    x: int
    y: int
    width: int
    height: int = 24
    emphasis: bool = False


class ContractPage(BaseModel):
    page: int
    title: str
    width: int
    height: int
    imageUrl: str | None = None
    imageLocalPath: str | None = None
    blocks: list[DocumentBlock] = Field(default_factory=list)
    evidences: list[EvidenceRef] = Field(default_factory=list)


class ContractSection(BaseModel):
    id: str
    title: str
    level: int
    page: int
    summary: str
    confidence: float
    sortOrder: int = 0
    sectionCode: str | None = None
    sectionPath: str | None = None
    blockIds: list[str] = Field(default_factory=list)
    evidenceId: str | None = None


class ClauseTag(BaseModel):
    id: str
    label: str
    coreLabel: str = "其他重要条款"
    labelSource: str = "core"
    title: str
    summary: str
    rawText: str
    page: int
    confidence: float
    sortOrder: int = 0
    sectionTitle: str | None = None
    references: list[str] = Field(default_factory=list)
    structuredFields: dict[str, Any] = Field(default_factory=dict)
    anchorText: str | None = None
    blockIds: list[str] = Field(default_factory=list)
    evidenceId: str
    needHumanReview: bool = False
    discoveryReason: str | None = None
    relatedAuditFocusIds: list[str] = Field(default_factory=list)


class KeyFact(BaseModel):
    id: str
    label: str
    value: str
    page: int
    confidence: float
    evidenceId: str | None = None
    notes: str | None = None


class ConfidenceOverview(BaseModel):
    overall: float
    sections: float
    clauses: float
    audit: float
    warnings: int


class KnowledgeBaseReviewStep(BaseModel):
    id: str
    label: str
    status: str = "pending"
    detail: str | None = None


class KnowledgeBaseReviewState(BaseModel):
    status: str = "idle"
    progressPercent: int = 0
    currentStepId: str | None = None
    currentStepLabel: str | None = None
    message: str | None = None
    detectedCategory: str | None = None
    matchedTemplateId: str | None = None
    matchedTemplateName: str | None = None
    issueCount: int | None = None
    steps: list[KnowledgeBaseReviewStep] = Field(default_factory=list)


class ContractTask(BaseModel):
    taskId: str
    fileName: str
    status: TaskStatus
    createdAt: str
    modelName: str
    confidenceOverview: ConfidenceOverview
    progressPercent: int = 0
    currentStage: str | None = None
    stageDetail: str | None = None
    elapsedMs: int = 0
    knowledgeBaseReview: KnowledgeBaseReviewState | None = None


class ContractAnalysisResult(BaseModel):
    task: ContractTask
    pages: list[ContractPage] = Field(default_factory=list)
    sections: list[ContractSection] = Field(default_factory=list)
    clauses: list[ClauseTag] = Field(default_factory=list)
    keyFacts: list[KeyFact] = Field(default_factory=list)


class UploadResponse(BaseModel):
    task_id: str

File: frontend/src/types/contract.ts
export type TaskStatus = "pending_upload" | "processing" | "completed" | "needs_review";
export type AnalysisTab = "sections" | "clauses" | "relations" | "audit" | "verification" | "knowledge" | "logs";

export interface EvidenceRef {
  id: string;
  page: number;
  bbox: [number, number, number, number];
  text: string;
  sourceType: "section" | "clause" | "fact";
  sourceId: string;
  segmentIndex: number;
  segmentCount: number;
  isPrimary: boolean;
  accent?: "cyan" | "amber";
}

export interface ContractSection {
  id: string;
  title: string;
  level: number;
  page: number;
  summary: string;
  confidence: number;
  sortOrder?: number;
  sectionCode?: string | null;
  sectionPath?: string | null;
  evidenceId?: string | null;
}

export interface ClauseTag {
  id: string;
  label: string;
  coreLabel: string;
  labelSource: "core" | "agent_discovered" | "user_configured";
  title: string;
  summary: string;
  rawText: string;
  page: number;
  confidence: number;
  sortOrder?: number;
  sectionTitle?: string | null;
  references?: string[];
  structuredFields?: Record<string, unknown>;
  anchorText?: string | null;
  evidenceId: string;
  needHumanReview: boolean;
  discoveryReason?: string | null;
  relatedAuditFocusIds: string[];
}

export interface DocumentBlock {
  id: string;
  text: string;
  x: number;
  y: number;
  width: number;
  height?: number;
  emphasis?: boolean;
}

export interface ContractPage {
  page: number;
  title: string;
  width: number;
  height: number;
  imageUrl?: string | null;
  blocks: DocumentBlock[];
  evidences: EvidenceRef[];
}

export interface KeyFact {
  id: string;
  label: string;
  value: string;
  page: number;
  confidence: number;
  evidenceId?: string | null;
  notes?: string | null;
}

export interface ConfidenceOverview {
  overall: number;
  sections: number;
  clauses: number;
  audit: number;
  warnings: number;
}

export interface KnowledgeBaseReviewStep {
  id: string;
  label: string;
  status: "pending" | "running" | "completed" | "failed";
  detail?: string | null;
}

export interface KnowledgeBaseReviewState {
  status: "idle" | "running" | "completed" | "failed";
  progressPercent: number;
  currentStepId?: string | null;
  currentStepLabel?: string | null;
  message?: string | null;
  detectedCategory?: string | null;
  matchedTemplateId?: string | null;
  matchedTemplateName?: string | null;
  issueCount?: number | null;
  steps: KnowledgeBaseReviewStep[];
}

export interface ContractTask {
  taskId: string;
  fileName: string;
  status: TaskStatus;
  createdAt: string;
  modelName: string;
  confidenceOverview: ConfidenceOverview;
  progressPercent: number;
  currentStage?: string | null;
  stageDetail?: string | null;
  elapsedMs?: number;
  knowledgeBaseReview?: KnowledgeBaseReviewState | null;
}

export interface ContractAnalysisResult {
  task: ContractTask;
  pages: ContractPage[];
  sections: ContractSection[];
  clauses: ClauseTag[];
  keyFacts: KeyFact[];
}
